std/bitmanip.d

```diff
@@ -886,6 +886,130 @@ public:
+                _ptr[fullWords] | endMask :
+                _ptr[fullWords] & ~endMask;
+    }
+
+    ///
+    @system unittest
+    {
+        import std.algorithm.comparison : equal;
+
+        auto b = BitArray([1, 0, 1, 0, 1, 1]);
+
+        b[] = true;
+        // all bits are set
+        assert(b.bitsSet.equal([0, 1, 2, 3, 4, 5]));
+
+        b[] = false;
+        // none of the bits are set
+        assert(b.bitsSet.empty);
+    }
+
+    /**
+      Sets the bits of a slice of $(D BitArray) which
+      starts at the index $(D start) and ends at the
+      index ($D end) - 1 with the values specified by $(D val).
+     */
+    void opSliceAssign(bool val, size_t start, size_t end)
+    in
+    {
+        assert(start >= 0 && start <= _len);
+        assert(end >=0 && end <= _len);
+        assert(start <= end);
+    }
+    body
+    {
+        size_t startBlock = start / bitsPerSizeT;
+        size_t endBlock = end / bitsPerSizeT;
+        size_t startOffset = start % bitsPerSizeT;
+        size_t endOffset = end % bitsPerSizeT;
+
+        if (startBlock == endBlock)
+        {
+            size_t startBlockMask = ~((size_t(1) << (startOffset)) - 1);
+            size_t endBlockMask = (size_t(1) << endOffset) - 1;
+            size_t joinMask = startBlockMask & endBlockMask;
+            _ptr[startBlock] = val ? _ptr[startBlock] | joinMask : _ptr[startBlock] & ~joinMask;
+            return;
+        }
+
+        size_t blockAdvance;
+        if (startOffset != 0)
+        {
+            size_t startBlockMask = ~((size_t(1) << (startOffset)) - 1);
+            _ptr[startBlock] = val ? _ptr[startBlock] | startBlockMask : _ptr[startBlock] & ~startBlockMask;
+            blockAdvance = 1;
+        }
+        if (endOffset != 0)
+        {
+            size_t endBlockMask = (size_t(1) << endOffset) - 1;
+            _ptr[endBlock] = val ? _ptr[endBlock] | endBlockMask : _ptr[endBlock] & ~endBlockMask;
+        }
+
+        for (size_t i = startBlock + blockAdvance; i < endBlock; i++)
+            _ptr[i] = val ? ~size_t(0) : 0;
+    }
+
+    ///
+    @system unittest
+    {
+        import std.algorithm.comparison : equal;
+        import std.range : iota;
+        import std.stdio;
+
+        auto b = BitArray([1, 0, 0, 0, 1, 1, 0]);
+        b[1 .. 3] = true;
+        assert(b.bitsSet.equal([0, 1, 2, 4, 5]));
+
+        bool[72] bitArray;
+        auto b1 = BitArray(bitArray);
+        b1[63 .. 67] = true;
+        assert(b1.bitsSet.equal([63, 64, 65, 66]));
+        b1[63 .. 67] = false;
+        assert(b1.bitsSet.empty);
+        b1[0 .. 64] = true;
+        assert(b1.bitsSet.equal(iota(0, 64)));
+
+        bool[256] bitArray2;
+        auto b2 = BitArray(bitArray2);
+        b2[3 .. 245] = true;
+        assert(b2.bitsSet.equal(iota(3, 245)));
+    }
+
+    /**
+      Flips all the bits in the $(D BitArray)
+     */
+    void flip()
+    {
+        foreach (i; 0 .. fullWords)
+            _ptr[i] = ~_ptr[i];
```

Comment:
the unittest employs a BitArray with just 6 bits. On my machine the size of a word is 64, so those are the endbits of a single word. I added an example which covers those lines.